doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+from 'normal' tables that are written to directly by users. The value of this will
+be one of `'direct'` or `'source'`.
+
+`data_source_id`: a nullable column that contains the `id` of the data source,
+if there is one. In the case of `CREATE TABLE .. FROM SOURCE` statements this would
+contain the `id` of the primary source.
+
+#### [mz_catalog.mz_sources](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_sources)
+
+`CREATE TABLE .. FROM SOURCE` statements will be added to this catalog table, despite
+them being 'tables', since this catalog table contains columns that are unique to
+source-related objects (e.g. `connection_id`).
+
+We will add the variant `'table'` to the `type` column which is currently a `text` column
+containing one of:
+`'kafka', 'mysql', 'postgres', 'load-generator', 'progress', or 'subsource'.`
+
+Once all customer's sources have been migrated to the new model, the `'subsource'` variant will be removed from the `type` column.
+
+Open Question: Will it be confusing to users that these source-fed tables will appear
+in both the `mz_sources` and `mz_tables` catalog tables?
+
+#### [mz_internal.mz_postgres_source_tables](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_postgres_source_tables) and [mz_internal.mz_mysql_source_tables](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_mysql_source_tables)
+
+Each row in these tables contains a mapping from a `subsource` `id` to the external
+`schema` and `table` of the Postgres/MySQL table being ingested into that subsource.
+
+These will be updated to include the same mappings for `CREATE TABLE .. FROM SOURCE`
+table objects to the external `schema` and `table` they are linked to.
+
+#### [mz_internal.mz_source_statistics](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_source_statistics)
+
+This table does not require any changes to its schema. However it is important to highlight
+that since multiple tables will be able to ingest from the same upstream external reference,
+the values reported in the following fields will be representative of the total number of
+records output to all tables rather than the number of rows ingested from the external references
+of the upstream system:
+
+-   updates_staged
+-   updates_committed
+-   records_indexed
+-   bytes_indexed
+-   snapshot_records_known
+-   snapshot_records_staged
+
+#### [mz_internal.mz_source_statuses](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_source_statuses)
+
+Similar to the change to `mz_catalog.mz_sources`, The `"table"` variant will be added
+to the `type` column and all `CREATE TABLE .. FROM SOURCE` statements
+will report a status in this table with the same semantics as existing subsources.
+
+#### [mz_internal.mz_source_status_history](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_source_status_history)
+
+No change is necessary to this table, besides noting that the `source_id` column will
```

Comment:
That's already what we do today for subsources, right? For Kafka, will we include both the source and the source table? Assuming that, unlike PostgreSQL and MySQL, these can't error independently.

Reply:
> That's already what we do today for subsources, right?

Yep

> For Kafka, will we include both the source and the source table? 

Yep we should such that kafka looks no different than the other source types